kernel/src/collections/ring_buffer.rs

```diff
@@ -318,3 +318,175 @@ mod test {
         assert_eq!(buf.dequeue(), None);
     }
 }
+
+// ===== Flux code ========
+// Below is a basic specification that checks:
+// 1. There are no division-by-zero errors
+// 2. There are no array bounds violations
+// 3. Some common-sense invariants (e.g., indexes into the RingBuffer are always valid)
+//
+// If any of these properties could be violated either in the RingBuffer implementation or how it
+// is used anywhere in Tock, Flux would raise an error.
+#[cfg(feature = "flux")]
+mod flux_specs {
+    // Prelude: Here we provide some specifications for methods/types in the core library.
+    // This allows Flux to make use of these specs for proving useful things about RingBuffer.
+    // Generally, these specs are per-project---if we verified many modules in Tock,
+    // there would only be one centralized set of these specs for all of Tock.
+    #[flux_rs::extern_spec]
+    impl<T> [T] {
+        // Need to tell Flux what slice.len() does
+        #[flux_rs::sig(fn(&[T][@len]) -> usize[len])]
+        fn len(v: &[T]) -> usize;
+
+        // Need to tell Flux what slice.split_at_mut() does (for our tests)
+        #[flux_rs::sig(fn(&mut [T][@len], usize[@mid]) -> (&mut [T][mid], &mut [T][len - mid]))]
+        fn split_at_mut(v: &mut [T], mid: usize) -> (&mut [T], &mut [T]);
+    }
+
+    // Need to tell Flux what an Option<T> is:
+    // Here, we refine Option<T> with a bool, denoting whether it is `Some` or `None`
+    #[flux_rs::extern_spec]
+    #[flux_rs::refined_by(b: bool)]
+    enum Option<T> {
+        #[variant(Option<T>[false])]
+        None,
+        #[variant({T} -> Option<T>[true])]
+        Some(T),
+    }
+
+    // ======= RingBuffer spec ===========
+    #[flux::specs {
+        mod collections {
+            mod ring_buffer {
+                // Specify well-formedness for RingBuffer<T>
+                //
+                // Flux will raise an error if
+                //  1) any RingBuffer implementation violates these rules,
+                //  2) any Tock code attempts to create a RingBuffer that
+                //     violates these rules.
```

Comment:
This is articulated kind of circularly---this is the spec, what it says is that if it is violated flux will error.

Instead, we want a comprehensible English description of what is specified and what the specification guarantees:

> A well-formed RingBuffer has an internal slice (`ring`) that has at least 1 element, and tracks the head and tail index, each must be less than the length of the internal slice. This ensures that an implementation can always use either `head` or `tail` to index into the slice.

(Check at least the last sentence for correctness)